rfcs/text/0000-rfc-library-preparation.md

```diff
@@ -0,0 +1,428 @@
+    - Cell suspension :left_right_arrow: Process_1 :left_right_arrow: Library preparation :left_right_arrow: Process_0 :left_right_arrow: Sequence file
+    - Process_1 :left_right_arrow: Library preparation protocol
+    - Process_0 :left_right_arrow: Sequencing protocol
+    - Declare rules about linking
+    
+---
+    
+![Figure 6](/rfcs/images/0000-lib_prep_rfc_fig6.png)
+
+**Figure 6: Diagram of how protocol linking will change with new library preparation entity**. A) Current metadata model showing that links to Sequencing protocol and Library preparation protocol entities are generated from the last process (arrow) in the graph. B) New metadata model showing that a link to the Sequencing protocol is generated from the ultimate process (process_0) in the graph, and a link to the Library preparation protocol is generated from the penultimate process (process_1) in the graph.
+
+---
+
+#### Data Storage Service
+
+- No changes will be needed.
+- Clients may need to update queries and subscriptions (some described below).
+
+#### Upload Service
+
+- No changes will be needed.
+
+#### Data Processing Pipelines
+
+- Subscription queries will reference a library preparation entity instead of cell suspension entity, where appropriate.
+- If multiple-bundles:1-workflow remains a use case, pipelines can query for all bundles with a particular library preparation UUID in order to trigger a single workflow.
+
+#### Data Brower
+
+- Is anything coded against cell suspension?
+- How will this affect displaying rows in Sample view?
+
+#### Matrix Service
+
+- Is anything coded against cell suspension?
+- Will there be disruptions to how Matrix Service identifies single cells if we insert a library preparation entity between cell suspension and sequence file (thinking especially for plate-based technologies)?
+
+#### Query Service
+
+- Is anything coded against cell suspension?
+- Will queries be made more difficult under this proposal?
+
+
+### Acceptance Criteria [optional]
+
+- Data processing pipelines workflows can successfully be started and run on all data files from the same library preparation.
+- Data consumers can find all data files associated with the same sequencing library preparation.
+- Data contributors can complete a metadata spreadsheet without much more difficulty than before (how is this measured?).
+
+### Unresolved Questions
+
+#### To solve during RFC process
+- How will the DCP enforce the requirement of having the library preparation entity in every logical unit (i.e. bundle)?
+- How will current production sequencing datasets be updated? All of the current datasets will need to be updated as all of them will need to have library preparation entities added. This is a complex update - change in "bundle" structure by adding a `library_preparation_0.json` entity - so it can not be done with simple AUDR. If the update happens after GA but before complex AUDR, the UUIDs will have to be maintained somehow.
```

Comment:
I think we must come up with a plan which doesn't require complete re-ingestion for this to work. There has been some internal discussions within ingest and we think this might be possible but it probably won't be pretty. 

Ensuring we are correctly representing the library prep strategy might need a discussion with some contributors but we should never need to discuss the bundle structure with a contributor

Reply:
I'd really prefer to take the time/resources required to make it pretty. A more ad hoc pseudo-re-ingestion strategy will introduce a lot of confusion to downstream components.

Reply:
> I assume there would need to be some bundle mergers that would require a conversation with the data contributor.

We have addressed all the merged bundles with the most recent re-ingestion. So, all the currently bundles are correct now. What would need to be changed is to insert a library preparation biomaterial JSON into each bundle, and update the linking to put the LP biomaterial JSON in the correct place in the experimental graph.